File: src/main/scala/robosky/uplands/client/render/blockentity/BlockEntityRenderingRegistry.scala
package robosky.uplands.client.render.blockentity

import net.fabricmc.fabric.api.client.rendereregistry.v1.BlockEntityRendererRegistry

import net.minecraft.block.entity.{BlockEntity, BlockEntityType}
import net.minecraft.client.render.block.entity.{BlockEntityRenderDispatcher, BlockEntityRenderer}

import robosky.uplands.block.bossroom.{ControlBlockEntity, DoorwayBlockEntity}

import scala.reflect.{ClassTag, classTag}

object BlockEntityRenderingRegistry {

  private def register[T <: BlockEntity](tpe: BlockEntityType[T], renderer: BlockEntityRenderDispatcher => BlockEntityRenderer[T]): Unit = {
    BlockEntityRendererRegistry.INSTANCE.register(tpe, renderer(_))
  }

  def init(): Unit = {
    register(ControlBlockEntity.TYPE, new ControlBlockEntityRenderer(_))
    register(DoorwayBlockEntity.TYPE, new DoorwayBlockEntityRenderer(_))
  }
}

File: src/main/java/robosky/uplands/client/render/blockentity/BlockEntityRenderingRegistry.java
package robosky.uplands.client.render.blockentity;

import java.util.function.Function;

import robosky.uplands.block.bossroom.ControlBlockEntity;
import robosky.uplands.block.bossroom.DoorwayBlockEntity;

import net.minecraft.block.entity.BlockEntity;
import net.minecraft.block.entity.BlockEntityType;
import net.minecraft.client.render.block.entity.BlockEntityRenderDispatcher;
import net.minecraft.client.render.block.entity.BlockEntityRenderer;

import net.fabricmc.fabric.api.client.rendereregistry.v1.BlockEntityRendererRegistry;

public final class BlockEntityRenderingRegistry {

    private static <T extends BlockEntity> void register(BlockEntityType<T> tpe, Function<BlockEntityRenderDispatcher, BlockEntityRenderer<T>> renderer) {
        BlockEntityRendererRegistry.INSTANCE.register(tpe, renderer);
    }

    public static void init() {
        register(ControlBlockEntity.TYPE, ControlBlockEntityRenderer::new);
        register(DoorwayBlockEntity.TYPE, DoorwayBlockEntityRenderer::new);
    }
}
